### PR TITLE
Add GIDAuthorizationFlowProcessor

### PR DESCRIPTION
**Pull request by Alex-4-Git**
GIDAuthorizationFlowProcessor can start, resume and cancel the authorization flow. This class is extracted out of `GIDSignIn.m` so GIDSignIn implementation can be simplified and easier to understand and test.

This PR adds protocol, implementation and unit tests. The fake will be added in the next PR.

### CODE REVIEW

File: GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/Implementations/GIDAuthorizationFlowProcessor.m [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+// Parameters for the auth and token exchange endpoints.
+static NSString *const kAudienceParameter = @"audience";
+
+static NSString *const kIncludeGrantedScopesParameter = @"include_granted_scopes";
+static NSString *const kLoginHintParameter = @"login_hint";
+static NSString *const kHostedDomainParameter = @"hd";
+
+@implementation GIDAuthorizationFlowProcessor {
+  // AppAuth external user-agent session state.
+  id<OIDExternalUserAgentSession> _currentAuthorizationFlow;
```
Comment (mdmathias) [inline GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/Implementations/GIDAuthorizationFlowProcessor.m:43]: If you add this to the protocol, then you could inject a fake in tests and wouldn't have to mock.

File: GoogleSignIn/Tests/Unit/GIDAuthorizationFlowProcessorTest.m [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+@end
+
+@implementation GIDAuthorizationFlowProcessorTest
+
+- (void)setUp {
+  [super setUp];
+  
+  _authorizationFlowProcessor = [[GIDAuthorizationFlowProcessor alloc] init];
+  _externalUserAgentSession = OCMProtocolMock(@protocol(OIDExternalUserAgentSession));
+  _authorizationServiceMock = OCMClassMock([OIDAuthorizationService class]);
```
Comment (mdmathias) [inline GoogleSignIn/Tests/Unit/GIDAuthorizationFlowProcessorTest.m:49]: If you keep these mocks, then I think you should be verifying things on them. I'm not sure you need to mock, or at least I don't think you need to mock `OIDExternalUserAgentSession`.

**Review by mdmathias — CHANGES_REQUESTED** [review]

File: GoogleSignIn/Tests/Unit/GIDAuthorizationFlowProcessorTest.m (hunk 2 above)
Comment (Alex-4-Git) [reply GoogleSignIn/Tests/Unit/GIDAuthorizationFlowProcessorTest.m:53]: Added OIDFakeExternalUserAgentSession.

File: GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/Implementations/GIDAuthorizationFlowProcessor.m (hunk 1 above)
Comment (Alex-4-Git) [reply GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/Implementations/GIDAuthorizationFlowProcessor.m]: Discussed offline. We decided to keep this as internal.

File: GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+/// The state of the authorization flow.
+@property(nonatomic, readonly, getter=isStarted) BOOL start;
+
+/// Starts the authorization flow.
+///
+/// @param options The @GIDSignInInternalOptions object to provide serverClientID, hostedDomain,
+///     clientID, scopes, loginHint and extraParams.
+/// @param emmSupport The EMM support info string.
+/// @param completion The block that is called on completion asynchronously.
```
[the remark this answers is not in the document]
Comment (mdmathias) [reply GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h:36]: I was suggesting that you add the explanation of the completion block's parameters to the doc comment for the completion param.

File: GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
```
Comment (petea) [inline GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h:19]: You can begin the NONNULL audited region below the `@class` forward declarations.

File: GoogleSignIn/Tests/Unit/GIDAuthorizationFlowProcessorTest.m (hunk 2 above)
Comment (mdmathias) [reply GoogleSignIn/Tests/Unit/GIDAuthorizationFlowProcessorTest.m:53]: Worth verifying the interactions you expect on the existing mock?

File: GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+@class OIDAuthorizationResponse;
+
+/// The protocol to control the authorization flow.
+@protocol GIDAuthorizationFlowProcessor <NSObject>
+
+/// The state of the authorization flow.
+@property(nonatomic, readonly, getter=isStarted) BOOL start;
+
+/// Starts the authorization flow.
+/// This method sends authorization request to AppAuth OIDAuthorizationService and gets back the response or an error.
```
Comment (petea) [inline GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h:31]: \`OIDAuthorizationService\`

Let's use backticks anytime we refer to symbols in a doc comment.

**Review by mdmathias — CHANGES_REQUESTED** [review]

File: GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+///
+/// @param options The @GIDSignInInternalOptions object to provide serverClientID, hostedDomain,
```
Comment (petea) [inline GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h:33]: \`GIDSignInInternalOptions\`

File: GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+///     clientID, scopes, loginHint and extraParams.
+/// @param emmSupport The EMM support info string.
+/// @param completion The block that is called on completion asynchronously.
+- (void)startWithOptions:(GIDSignInInternalOptions *)options
+              emmSupport:(nullable NSString *)emmSupport
+              completion:(void (^)(OIDAuthorizationResponse *_Nullable authorizationResponse,
+                                   NSError *_Nullable error))completion;
+
+/// Handles the custom scheme URL opened by SFSafariViewController to cleanup UI on iOS 10.
```
Comment (petea) [inline GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h:42]: `custom URL scheme`

Comment (petea) [inline GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h:42]: Let's not forget the Device Policy app and the EMM case.  Also, this path is returning control to the 3P app, not sure what "cleanup UI" is referring to.

**Review by petea — CHANGES_REQUESTED** [review]

File: GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h (hunk 4 above)
Comment (Alex-4-Git) [reply GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h]: Done

File: GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h (hunk 5 above)
Comment (Alex-4-Git) [reply GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h]: Done.

File: GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h (hunk 6 above)
Comment (Alex-4-Git) [reply GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h]: Done.

File: GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (Alex-4-Git) [reply GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h:37]: Done.

File: GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h (hunk 7 above)
Comment (Alex-4-Git) [reply GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h]: Done.

Comment (Alex-4-Git) [reply GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h]: In GIDSignIn code the Device Policy app is handled by this this method. This is only for custom custom URL.

Comment (Alex-4-Git) [reply GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h]: Removed `cleanup UI`.

File: GoogleSignIn/Tests/Unit/GIDAuthorizationFlowProcessorTest.m (hunk 2 above)
Comment (Alex-4-Git) [reply GoogleSignIn/Tests/Unit/GIDAuthorizationFlowProcessorTest.m:54]: Added more obvious verification.

**Review by mdmathias — APPROVED** [review]

**Review by petea — CHANGES_REQUESTED** [review]

File: .swiftpm/xcode/xcshareddata/xcschemes/GoogleSignIn-UnitTests.xcscheme [hunk 8]
```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (Alex-4-Git) [reply .swiftpm/xcode/xcshareddata/xcschemes/GoogleSignIn-UnitTests.xcscheme]: It's a mistake.

File: GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+NS_ASSUME_NONNULL_BEGIN
+
+/// The protocol to control the authorization flow.
+@protocol GIDAuthorizationFlowProcessor <NSObject>
+
+/// The state of the authorization flow.
+@property(nonatomic, readonly, getter=isStarted) BOOL start;
+
+/// Starts the authorization flow.
```
[the remark this answers is not in the document]
Comment (Alex-4-Git) [reply GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h:30]: Done.

File: GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/Implementations/GIDAuthorizationFlowProcessor.h [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+#import <Foundation/Foundation.h>
+
+#import "GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/API/GIDAuthorizationFlowProcessor.h"
+
+@class OIDServiceConfiguration;
+
+NS_ASSUME_NONNULL_BEGIN
+
+/// Concrete implementation of the protocol GIDAuthorizationFlowProcessor.
```
[the remark this answers is not in the document]
Comment (Alex-4-Git) [reply GoogleSignIn/Sources/GIDAuthorizationFlowProcessor/Implementations/GIDAuthorizationFlowProcessor.h]: Done.

**Pull request merged**